mrjob/fs/gcs.py

```diff
@@ -0,0 +1,224 @@
+                    continue
+                yield uri, long(item["size"]), item["bucket"], item["name"], item["md5Hash"], item["etag"]
+            req = self.service.objects().list_next(req, resp)
+
+    def md5sum(self, path):
+        object_list = list(self.lsWithInfo(path))
+        print object_list
+        if len(object_list) == 1:
+            return binascii.hexlify(base64.decodestring(object_list[0][4]))
+        else:
+            raise Exception("path for md5 sum doesn't resolve to single object" + path)
+
+    def _cat_file(self, filename):
+        bucket_name, object_name = parse_gcs_uri(filename)
+        req = self.service.objects().get_media(bucket=bucket_name, object=object_name)
+        fh = io.BytesIO()
+        downloader = MediaIoBaseDownload(fh, req, chunksize=1024*1024)
+        done = False
+        while not done:
+            status, done = downloader.next_chunk()
+            if status:
+                log.info("Download %d%%." % int(status.progress() * 100))
+            log.info("Download Complete for " + filename)
+        """ TODO: So, at this point we have the file content as you can see it using fh.getvalue().
+        But rather than re-implementing the decompression logic (to deal with compressed files) I'd rather
+        use the same read_file() method (in util.py) that the S3 and local versions are using.
+        But I can't figure out how to build the right file object to pass it to read_file().
+        Help appreciated.
+        """
+
+    def mkdir(self, dest):
+        """Make a directory. This does nothing on GCS because there are
+        no directories.
+        """
+        raise NotImplementedError
+
+    def exists(self, path_glob):
+        """Does the given path exist?
+
+        If dest is a directory (ends with a "/"), we check if there are
+        any files starting with that path.
+        """
+        # just fall back on ls(); it's smart
+        try:
+            paths = self.ls(path_glob)
+        except:
+            paths = []
+        return any(paths)
+
+    def rm(self, path_glob):
+        """Remove all files matching the given glob."""
+        for item in self.lsWithInfo(path_glob):
+            req = self.service.objects().delete(bucket=item[2], object=item[3])
+            log.debug("deleting " + item[0])
+            req.execute()
+
+    def touchz(self, dest):
+        """ TODO: too much common code between touchz and updload. Refactor.
+        Make an empty file in the given location. Raises an error if
+        a file already exists in that location."""
+        if self.exists(dest):
+            raise Exception("File already exists: " + dest)
+        media = http.MediaIoBaseUpload(io.BytesIO(""), "text/plain")
+        req = self.service.objects().insert( bucket=parse_gcs_uri(dest)[0], name=parse_gcs_uri(dest)[1], media_body=media)
+        req.execute()
+
+    def upload(self, local_path, destination_uri):
```

Comment:
Yeah, this really should be part of the `Filesystem` API. But it should be named `put` (like the `hadoop fs` command).